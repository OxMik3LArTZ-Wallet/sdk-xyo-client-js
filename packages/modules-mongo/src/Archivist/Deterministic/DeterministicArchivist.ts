import { ArchivistConfigSchema } from '@xyo-network/archivist'
import { MongoDBArchivist, MongoDBArchivistConfig, MongoDBArchivistConfigSchema, MongoDBArchivistParams } from '@xyo-network/archivist-mongodb'

export const MongoDBDeterministicArchivistConfigSchema = MongoDBArchivistConfigSchema
export type MongoDBDeterministicArchivistConfigSchema = typeof MongoDBDeterministicArchivistConfigSchema

export type MongoDBDeterministicArchivistConfig = MongoDBArchivistConfig

export type MongoDBDeterministicArchivistParams = MongoDBArchivistParams

interface Labels<T extends Record<string, string>> {
  labels: T
}

interface MongoStorageClass {
  storageClass: 'mongodb'
}

export class MongoDBDeterministicArchivist<
  TParams extends MongoDBDeterministicArchivistParams = MongoDBDeterministicArchivistParams,
> extends MongoDBArchivist<TParams> {
  static override configSchemas = [MongoDBDeterministicArchivistConfigSchema, MongoDBArchivistConfigSchema, ArchivistConfigSchema]
}
