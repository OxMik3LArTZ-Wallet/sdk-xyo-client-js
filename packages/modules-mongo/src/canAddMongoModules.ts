import { hasMongoDBConfig } from './Mongo'

export const canAddMongoModules = (): boolean => {
  return hasMongoDBConfig()
}
