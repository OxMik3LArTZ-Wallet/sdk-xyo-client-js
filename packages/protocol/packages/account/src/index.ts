export * from './Account'
export * from './HDAccount'
export * from './HDWallet'
export * from './Key'
