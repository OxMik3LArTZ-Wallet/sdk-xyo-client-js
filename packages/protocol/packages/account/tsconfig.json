{
  "compilerOptions": {
    "composite": true,
    "declarationDir": "./dist/types",
    "experimentalDecorators": true,
    "typeRoots": []
  },
  "exclude": ["dist"],
  "extends": "@xylabs/tsconfig"
}

