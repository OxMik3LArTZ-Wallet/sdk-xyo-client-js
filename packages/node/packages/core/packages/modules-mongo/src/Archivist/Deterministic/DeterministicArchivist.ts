import { assertEx } from '@xylabs/assert'
import { exists } from '@xylabs/exists'
import { fulfilledValues } from '@xylabs/promise'
import { Account } from '@xyo-network/account'
import {
  AbstractArchivist,
  ArchivistConfig,
  ArchivistConfigSchema,
  ArchivistGetQuery,
  ArchivistGetQuerySchema,
  ArchivistInsertQuery,
  ArchivistInsertQuerySchema,
  ArchivistParams,
  ArchivistQuery,
} from '@xyo-network/archivist'
import { BoundWitness } from '@xyo-network/boundwitness-model'
import { BoundWitnessWrapper } from '@xyo-network/boundwitness-wrapper'
import { handleErrorAsync } from '@xyo-network/error'
import {
  AnyConfigSchema,
  ModuleConfig,
  ModuleError,
  ModuleErrorBuilder,
  ModuleQueryResult,
  QueryBoundWitness,
  QueryBoundWitnessWrapper,
} from '@xyo-network/module'
import { BoundWitnessWithMeta, PayloadWithMeta, PayloadWithPartialMeta } from '@xyo-network/node-core-model'
import { Payload } from '@xyo-network/payload-model'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import { BaseMongoSdk } from '@xyo-network/sdk-xyo-mongo-js'

import { validByType } from './validByType'

export type MongoDBDeterministicArchivistParams = ArchivistParams<
  AnyConfigSchema<ArchivistConfig>,
  {
    boundWitnessSdk: BaseMongoSdk<BoundWitnessWithMeta>
    payloadSdk: BaseMongoSdk<PayloadWithMeta>
  }
>

const toBoundWitnessWithMeta = async (wrapper: BoundWitnessWrapper | QueryBoundWitnessWrapper): Promise<BoundWitnessWithMeta> => {
  const bw = wrapper.boundwitness as BoundWitness
  return { ...bw, _hash: await wrapper.hashAsync(), _timestamp: Date.now() }
}

const toReturnValue = (value: Payload | BoundWitness): Payload => {
  const _signatures = (value as BoundWitness)?._signatures
  if (_signatures) {
    return { ...PayloadWrapper.wrap(value).body(), _signatures } as BoundWitness
  } else {
    return { ...PayloadWrapper.wrap(value).body() }
  }
}

const toPayloadWithMeta = async (wrapper: PayloadWrapper): Promise<PayloadWithMeta> => {
  return { ...wrapper.payload(), _hash: await wrapper.hashAsync(), _timestamp: Date.now() }
}

export class MongoDBDeterministicArchivist<
  TParams extends MongoDBDeterministicArchivistParams = MongoDBDeterministicArchivistParams,
> extends AbstractArchivist<TParams> {
  static override configSchemas = [ArchivistConfigSchema]

  get boundWitnesses() {
    return this.params.boundWitnessSdk
  }

  get payloads() {
    return this.params.payloadSdk
  }

  override get queries(): string[] {
    return [ArchivistInsertQuerySchema, ...super.queries]
  }

  override get(_hashes: string[]): Promise<Payload[]> {
    throw new Error('get method must be called via query')
  }

  override async head(): Promise<Payload | undefined> {
    const head = await (await this.payloads.find({})).sort({ _timestamp: -1 }).limit(1).toArray()
    return head[0] ? PayloadWrapper.wrap(head[0]).body() : undefined
  }

  insert(_items: Payload[]): Promise<BoundWitness[]> {
    throw new Error('insert method must be called via query')
  }

  protected async getInternal(wrapper: QueryBoundWitnessWrapper<ArchivistQuery>, queryPayload: ArchivistGetQuery): Promise<Payload[]> {
    const hashes = queryPayload.hashes
    const payloads = hashes.map((_hash) => this.payloads.findOne({ _hash }))
    const bws = hashes.map((_hash) => this.boundWitnesses.findOne({ _hash }))
    const gets = await Promise.allSettled([payloads, bws].flat())
    const succeeded = gets.reduce<(PayloadWithPartialMeta | null)[]>(fulfilledValues, [])
    return succeeded.filter(exists).map(toReturnValue)
  }

  protected async insertInternal(wrapper: QueryBoundWitnessWrapper<ArchivistQuery>, queryPayload: ArchivistInsertQuery): Promise<BoundWitness[]> {
    const toStore = [wrapper.boundwitness, ...wrapper.payloadsArray.map((p) => p.payload())]
    const [bw, p] = await validByType(toStore)
    const boundWitnesses = await Promise.all(bw.map((x) => toBoundWitnessWithMeta(x)))
    const payloads = await Promise.all(p.map((x) => toPayloadWithMeta(x)))
    if (boundWitnesses.length) {
      const boundWitnessesResult = await this.boundWitnesses.insertMany(boundWitnesses)
      if (!boundWitnessesResult.acknowledged || boundWitnessesResult.insertedCount !== boundWitnesses.length)
        throw new Error('MongoDBDeterministicArchivist: Error inserting BoundWitnesses')
    }
    if (payloads.length) {
      const payloadsResult = await this.payloads.insertMany(payloads)
      if (!payloadsResult.acknowledged || payloadsResult.insertedCount !== payloads.length)
        throw new Error('MongoDBDeterministicArchivist: Error inserting Payloads')
    }
    const [result] = await this.bindQueryResult(queryPayload, [wrapper.boundwitness, ...wrapper.payloadsArray.map((p) => p.payload())])
    return [result[0]]
  }

  protected override async queryHandler<T extends QueryBoundWitness = QueryBoundWitness, TConfig extends ModuleConfig = ModuleConfig>(
    query: T,
    payloads?: Payload[],
    queryConfig?: TConfig,
  ): Promise<ModuleQueryResult> {
    const wrapper = QueryBoundWitnessWrapper.parseQuery<ArchivistQuery>(query, payloads)
    const queryPayload = await wrapper.getQuery()
    assertEx(this.queryable(query, payloads, queryConfig))
    const resultPayloads: Payload[] = []
    const errorPayloads: ModuleError[] = []
    // TODO: Use new Account once we mock Account.new in Jest
    const queryAccount = Account.randomSync()
    // const queryAccount = Account.randomSync()
    try {
      switch (queryPayload.schema) {
        case ArchivistGetQuerySchema: {
          resultPayloads.push(...(await this.getInternal(wrapper, queryPayload)))
          break
        }
        case ArchivistInsertQuerySchema: {
          resultPayloads.push(...(await this.insertInternal(wrapper, queryPayload)))
          break
        }
        default:
          return super.queryHandler(query, payloads)
      }
    } catch (ex) {
      await handleErrorAsync(ex, async (error) => {
        errorPayloads.push(
          new ModuleErrorBuilder()
            .sources([await wrapper.hashAsync()])
            .name(this.config.name ?? '<Unknown>')
            .query(query.schema)
            .message(error.message)
            .build(),
        )
      })
    }
    return (await this.bindQueryResult(queryPayload, resultPayloads, [queryAccount], errorPayloads))[0]
  }
}
