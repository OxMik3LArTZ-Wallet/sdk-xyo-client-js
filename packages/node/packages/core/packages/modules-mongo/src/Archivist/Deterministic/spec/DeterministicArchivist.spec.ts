/* eslint-disable max-statements */
let timestamp = 123456789000
Date.now = jest.fn(() => timestamp)

import { describeIf } from '@xylabs/jest-helpers'
import { Account } from '@xyo-network/account'
import { ArchivistWrapper } from '@xyo-network/archivist'
import { BoundWitnessBuilder } from '@xyo-network/boundwitness-builder'
import { BoundWitness } from '@xyo-network/boundwitness-model'
import { BoundWitnessWrapper } from '@xyo-network/boundwitness-wrapper'
import { BoundWitnessWithMeta, PayloadWithMeta } from '@xyo-network/node-core-model'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import { BaseMongoSdk, BaseMongoSdkConfig } from '@xyo-network/sdk-xyo-mongo-js'

import { canAddMongoModules } from '../../../canAddMongoModules'
import { COLLECTIONS } from '../../../collections'
import { MongoDBDeterministicArchivist } from '../DeterministicArchivist'

describeIf(canAddMongoModules())('DeterministicArchivist', () => {
  const boundWitnessesConfig: BaseMongoSdkConfig = { collection: COLLECTIONS.BoundWitnesses }
  const payloadsConfig: BaseMongoSdkConfig = { collection: COLLECTIONS.Payloads }
  const archiveAccount: Account = new Account({ phrase: 'temp' })
  // 0x10cal
  const userAccount: Account = new Account({ privateKey: '69f0b123c094c34191f22c25426036d6e46d5e1fab0a04a164b3c1c2621152ab' })
  // 0xdadda
  const moduleAccount: Account = new Account({ privateKey: '9c9637dc07ce9956190c028677f5195a8fb425e9927bf2e48fe39a1c55cf050a' })
  // 0xace
  const randomAccount: Account = new Account({ privateKey: '3c17e038c8daeed7dfab9b9653321523d5f1a68eadfc5e4bd501075a5e43bbcc' })
  const payload1 = { nonce: 1, schema: 'network.xyo.debug' }
  const payload2 = { nonce: 2, schema: 'network.xyo.test' }
  const payload3 = { nonce: 3, schema: 'network.xyo.debug' }
  const payload4 = { nonce: 4, schema: 'network.xyo.test' }
  const payloadWrapper1 = PayloadWrapper.parse(payload1)
  const payloadWrapper2 = PayloadWrapper.parse(payload2)
  const payloadWrapper3 = PayloadWrapper.parse(payload3)
  const payloadWrapper4 = PayloadWrapper.parse(payload4)
  const boundWitness1 = new BoundWitnessBuilder().payload(payloadWrapper1.payload).witness(userAccount).build()[0]
  const boundWitness2 = new BoundWitnessBuilder().payload(payloadWrapper2.payload).witness(userAccount).build()[0]
  const boundWitness3 = new BoundWitnessBuilder().payloads([payloadWrapper3.payload, payloadWrapper4.payload]).witness(userAccount).build()[0]
  const boundWitnessWrapper1 = BoundWitnessWrapper.parse(boundWitness1)
  boundWitnessWrapper1.payloads = [payload1]
  const boundWitnessWrapper2 = BoundWitnessWrapper.parse(boundWitness2)
  boundWitnessWrapper2.payloads = [payload2]
  const boundWitnessWrapper3 = BoundWitnessWrapper.parse(boundWitness3)
  boundWitnessWrapper3.payloads = [payload3, payload4]

  let archivist: ArchivistWrapper
  let insertResult1: BoundWitness[]
  // let insertResult2: BoundWitness[]
  let insertResult3: BoundWitness[]
  const insertResults: BoundWitness[][] = []
  beforeAll(async () => {
    jest.spyOn(Account, 'random').mockImplementation(() => randomAccount)
    boundWitnessesConfig.dbConnectionString = process.env.MONGO_CONNECTION_STRING
    payloadsConfig.dbConnectionString = process.env.MONGO_CONNECTION_STRING
    const boundWitnesses: BaseMongoSdk<BoundWitnessWithMeta> = new BaseMongoSdk(boundWitnessesConfig)
    const payloads: BaseMongoSdk<PayloadWithMeta> = new BaseMongoSdk(payloadsConfig)
    const module = await MongoDBDeterministicArchivist.create({
      account: moduleAccount,
      boundWitnessSdk: boundWitnesses,
      config: { schema: MongoDBDeterministicArchivist.configSchema },
      payloadSdk: payloads,
    })
    archivist = ArchivistWrapper.wrap(module, archiveAccount)
    const insertions = [
      // TODO: Try simple cases of [payload, BW, mixed BW & Payload]
      [boundWitness1, payload1],
      [boundWitness2, payload2],
      [boundWitness3, payload3, payload4],
    ]
    for (const insertion of insertions) {
      const insertionResult = await archivist.insert(insertion)
      insertResults.push(insertionResult)
      // NOTE: Increment Date.now after each insert so that DB sorting by time works
      timestamp++
    }
    insertResult1 = insertResults[0]
    // insertResult2 = insertResults[1]
    insertResult3 = insertResults[2]
  })
  describe('discover', () => {
    it('discovers module', async () => {
      const result = await archivist.discover()
      expect(result).toBeArray()
      expect(result.length).toBeGreaterThan(0)
    })
  })
  describe('insert', () => {
    it('inserts single payload', () => {
      expect(insertResult1).toBeTruthy()
      expect(insertResult1).toBeArrayOfSize(2)
      const [boundResult, transactionResults] = insertResult1
      expect(boundResult.addresses).toContain(archivist.address)
      expect(transactionResults.addresses).toContain(moduleAccount.public.address.hex)
      expect(transactionResults.payload_hashes).toBeArrayOfSize(boundWitnessWrapper1.payloadsArray.length + 3)
      boundWitnessWrapper1.payloadsArray.forEach((p) => {
        expect(transactionResults.payload_hashes).toInclude(p.hash)
      })
      expect(insertResult1.map((bw) => BoundWitnessWrapper.parse(bw).boundwitness)).toMatchSnapshot()
    })
    it('inserts multiple payloads', () => {
      expect(insertResult3).toBeTruthy()
      expect(insertResult3).toBeArrayOfSize(2)
      const [boundResult, transactionResults] = insertResult3
      expect(boundResult.addresses).toContain(archivist.address)
      expect(transactionResults.addresses).toContain(moduleAccount.public.address.hex)
      expect(transactionResults.payload_hashes).toBeArrayOfSize(boundWitnessWrapper3.payloadsArray.length + 3)
      boundWitnessWrapper3.payloadsArray.forEach((p) => {
        expect(transactionResults.payload_hashes).toInclude(p.hash)
      })
      expect(insertResult3.map((bw) => BoundWitnessWrapper.parse(bw).boundwitness)).toMatchSnapshot()
    })
  })
  describe('get', () => {
    it.each([
      ['gets single payload', [payloadWrapper1]],
      ['gets multiple payloads', [payloadWrapper1, payloadWrapper2, payloadWrapper3]],
      ['gets single boundwitness', [boundWitnessWrapper1]],
      ['gets multiple boundwitness', [boundWitnessWrapper1, boundWitnessWrapper2, boundWitnessWrapper3]],
    ])('%s', async (_title, payloads) => {
      const results = await archivist.get(payloads.map((p) => p.hash))
      expect(results).toBeTruthy()
      expect(results).toBeArrayOfSize(payloads.length)
      const resultPayloads = results.map((result) => PayloadWrapper.parse(result))
      const resultHashes = resultPayloads.map((p) => p.hash)
      payloads.map((p) => {
        expect(resultHashes).toInclude(p.hash)
      })
      expect(results).toMatchSnapshot()
    })
  })
})
