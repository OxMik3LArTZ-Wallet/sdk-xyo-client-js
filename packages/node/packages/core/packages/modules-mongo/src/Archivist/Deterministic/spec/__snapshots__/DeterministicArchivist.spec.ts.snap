// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeterministicArchivist get gets multiple boundwitness 1`] = `
[
  {
    "_signatures": [
      "325501822063bf69e34f6a64f96ef1e55e8eed5060c0dc300271be82b4df024163b2b294018c5d06df160b20394f460b955c77d8a5147ec7f1bafa79b4798662",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "_signatures": [
      "1810b9f9c6623adf38d0093ea6f3c2125570ed1ee757dbe9e9221b4ffb3920ec0b11175fdffc7e0bfee5d844f9e43c7e261d2bfcd1b273b46caa5cb23a4c798d",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "1c9033e467d6877e2c84a70b29212a585bd6fd448bccfcc12adf0505c18984f6",
    ],
    "payload_schemas": [
      "network.xyo.test",
    ],
    "previous_hashes": [
      "58c138dd2aab7bd78bd2c3d80bebf900d3f99f506e201f1f0ace1d473fc45d76",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "_signatures": [
      "29a8dfe2cf7675e8c371763d6d8c7d9b0990145191967c6e399fff7854cd22390d6a8a7f4ae7628ac62deea6c34af33c37623950b1aba6ae4b8c88683d615c20",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "c5e4a2513ea22a7ccfa359207b4d83af68688dca8c9a3f5d39106ba6e599b481",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist get gets multiple payloads 1`] = `
[
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
  {
    "nonce": 2,
    "schema": "network.xyo.test",
  },
  {
    "nonce": 3,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist get gets single boundwitness 1`] = `
[
  {
    "_signatures": [
      "325501822063bf69e34f6a64f96ef1e55e8eed5060c0dc300271be82b4df024163b2b294018c5d06df160b20394f460b955c77d8a5147ec7f1bafa79b4798662",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist get gets single payload 1`] = `
[
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist insert inserts multiple payloads 1`] = `
[
  {
    "_signatures": [
      "7a91a914b66a78391a69eb211c57097de9bfc1dd0c9c2d80d6f54b131b872973611f84c9f369d6e6f290e13e1eb261a20b895e22216fd719244df32a80643353",
      "895a38c46575b89279d6445916dc70eac421126322cb46e82c55e62c60c0265b3af9a1818725d4dcb743c80b72b0c351326fce66521db57e2c6ad1db02d5637c",
    ],
    "addresses": [
      "daddab0e0468c920bd5aff4b14fd94c20a598055",
      "babe1d55e51844ea1cdc6b4dcbb649bb08e3cc3c",
    ],
    "payload_hashes": [
      "f3b73fb620a239b3c900d0a44ef0bf2bb6a8418054edab5cef41a849042f5795",
    ],
    "payload_schemas": [
      "network.xyo.boundwitness",
    ],
    "previous_hashes": [
      "f3b73fb620a239b3c900d0a44ef0bf2bb6a8418054edab5cef41a849042f5795",
      "7b888e7b0ee8bf2413498e7169321fa9e975eaffbb1b2e65a97f439ad7318244",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789002,
  },
  {
    "_signatures": [
      "ed8e4f25fe06f04e6fe18c7a5d1a7573ae292a992e0418962bfe933b6c80fdbe15c046cfeffb91e418cbb01e189805f07d4fc8b3090bb4096c9943195b7d6e57",
    ],
    "addresses": [
      "daddab0e0468c920bd5aff4b14fd94c20a598055",
    ],
    "payload_hashes": [
      "d0796172dfb3f66f62e69b1120ef10c6e767e8b260272da5802865e7add48bcf",
      "4e2d335e7f54334b10e30706546f820a750974381586ec9aa2eca6f916249b25",
      "e307ab9ba8ea876135ae22ff47114513cf7fefe024dab21d89d8584378df6b11",
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.boundwitness",
      "network.xyo.query.archivist.insert",
      "network.xyo.boundwitness",
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "7b888e7b0ee8bf2413498e7169321fa9e975eaffbb1b2e65a97f439ad7318244",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789002,
  },
]
`;

exports[`DeterministicArchivist insert inserts single payload 1`] = `
[
  {
    "_signatures": [
      "ace9c243e6e206a1e2f73598606d3c9342ffd94db87e5cfce857379836e8d5a67018286c97e74f0557ec16c551d3a54da6640254aed90aa7e94e41aceae82c0e",
      "e4aa358e52fb6a166810d1f60d523617a434f38488e5f696c8d6bb5388c01ce7578f5efe74eb60168b3de4bd3cd2ee002daed1f0c4355ccad710c777895e313b",
    ],
    "addresses": [
      "daddab0e0468c920bd5aff4b14fd94c20a598055",
      "babe1d55e51844ea1cdc6b4dcbb649bb08e3cc3c",
    ],
    "payload_hashes": [
      "0b625a70ca20763ed5f1557358be488ede245cc90724a09b6aead493da0220a7",
    ],
    "payload_schemas": [
      "network.xyo.boundwitness",
    ],
    "previous_hashes": [
      "0b625a70ca20763ed5f1557358be488ede245cc90724a09b6aead493da0220a7",
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "_signatures": [
      "2d357722f9e0f03e932c7c940cc38c57b78c207ac8634fe1dd643c3e9bc346255e5dbb5ad66e2538d529305d20b92febcbb44a443033c5c91926c95bdda7bae4",
    ],
    "addresses": [
      "daddab0e0468c920bd5aff4b14fd94c20a598055",
    ],
    "payload_hashes": [
      "860b201b9d8891de3917ca7bfc5a4ec8898f6ff36b3ded59c9b279be38a45f78",
      "81b707afc5742a5501a4fac4ad56132928ce711de36de7aeda936e08066cfcad",
      "58c138dd2aab7bd78bd2c3d80bebf900d3f99f506e201f1f0ace1d473fc45d76",
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.boundwitness",
      "network.xyo.query.archivist.insert",
      "network.xyo.boundwitness",
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;
